### PR TITLE
A React-based PWA for displaying AMP content

### PR DESCRIPTION
**Pull request by choumx**
A simple React-based [progressive web app](https://addyosmani.com/blog/getting-started-with-progressive-web-apps/) that displays [Accelerated Mobile Page (AMP)](https://ampproject.org) content. Built on [create-react-app](https://github.com/facebookincubator/create-react-app) for minimal build configuration.

Live demo: http://choumx.github.io/amp-pwa

This project adds modern web features to `create-react-app`:

* **Accelerated mobile page (AMP) content:** Displays fast-loading AMP documents within the app shell via Shadow DOM.
* **Progressive web:** A service worker enables progressive enhancement of AMP content with precaching, offline functionality and an app shell.

Compared to `create-react-app`, this project adds a small number of new dependencies:

- `sw-precache` generates a production-ready service worker for precaching and other progressive enhancements.
- `express` is used to run the development API server, `server.js`.
- `bootstrap` and `react-bootstrap` can be easily removed for your choice of UI framework.

### CODE REVIEW

**Comment by choumx** [comment]
@cramforce PTAL

File: amp-pwa/manifest.webmanifest [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  "background_color": "#fff",
+  "description": "A simple React-based progressive web app that displays Accelerated Mobile Page (AMP) content.",
+  "icons": [
+    {
+      "src": "/icon.png",
+      "sizes": "192x192",
+      "type": "image/png"
+    }
+  ]
+}
```
Comment (cramforce) [inline amp-pwa/manifest.webmanifest]: Add trailing newlines here and elsewhere.

File: amp-pwa/src/components/home.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+            <Article title={doc.title} subtitle={doc.subtitle} src={doc.url} key={doc.url} />
+          )}
+        </div>
+      </div>
+    );
+  }
+}
+Home.propTypes = {
+  documents: React.PropTypes.arrayOf(React.PropTypes.object),
+}
```
Comment (cramforce) [inline amp-pwa/src/components/home.js]: `;` :)

**Comment by sebastianbenz** [comment]
The demo doesn't work on safari - if this is intended, we should mention it in the limitations.

**Comment by cramforce** [comment]
Could you fix the images?

When I'm going to a deep link directly it serves the AMP doc (yaihh), but then SW installation fails with a 404.

**Comment by dvoytenko** [comment]
@cramforce I noticed errors yesterday and fixed in a separate PR. Unfortunately, this part broken on resources refactoring. We'll need to setup integration tests for this part.

File: amp-pwa/manifest.webmanifest (hunk 1 above)
Comment (choumx) [inline amp-pwa/manifest.webmanifest]: Done.

File: amp-pwa/src/components/home.js (hunk 2 above)
Comment (choumx) [inline amp-pwa/src/components/home.js]: Done.

**Comment by choumx** [comment]
> The demo doesn't work on safari - if this is intended, we should mention it in the limitations.

@sebastianbenz Not intended -- I'm investigating a problem in Safari with the minimized version of the AMP-shadow runtime, `shadow-v0.js`. My latest build bundles the unminimized version via Webpack and it works in Safari: http://choumx.github.io/amp-pwa

> Could you fix the images?

Done. I also worked around the shadow bug by bundling a local build, although it breaks `amp-social-share` and `amp-list` due to version mismatch with extensions. It works better than before, though hopefully everything will work in the next prod push.

> When I'm going to a deep link directly it serves the AMP doc (yaihh), but then SW installation fails with a 404.

That was probably the `amp-install-serviceworker` extension in each content page. I've removed those as well as some unnecessary UI.

**Comment by cramforce** [comment]
Can you make URLs like
https://choumx.github.io/amp-pwa/content/housing.amp.html hit the SW and
load the shell when they get loaded with SW installed?

**Comment by choumx** [comment]
That should already work and I just tested it again on desktop Chrome. If you see the nav bar at the top, then the shell was rendered.

**Comment by cramforce** [comment]
Yeah, never mind. It is working now.

**Comment by cramforce** [comment]
One request, though: The leaf pages should use amp-install-serviceworker to
bootstrap the app.

**Comment by sebastianbenz** [comment]
Few more comments from my side:

* the link to the homepage is not working (amp-pwa) is missing
* can we include cross references between articles to demonstrate how the PWA handles these?
* I would pick different sample pages:
  * The hello world sample article only shows the AMP icon
  * The live blog sample doesn't work
  * The search in the product sample leads back to ampbyexample.com

**Comment by cramforce** [comment]
@sebastianbenz We're looking at actually getting good newsy-content soon. The ABE stuff is just to have something at all.

**Comment by choumx** [comment]
> The leaf pages should use amp-install-serviceworker to
bootstrap the app.

Done.

> the link to the homepage is not working (amp-pwa) is missing

Fixed, thanks!

> can we include cross references between articles to demonstrate how the PWA handles these?

@sebastianbenz Could you elaborate on what you mean by this? E.g. have the list of articles in the nav bar, or in the AMP documents themselves?

**Comment by cramforce** [comment]
@chmoux I think what @sebastianbenz meant is something like this:

`<a href="/link/to/article.html">Article</a>`

Clicking this should trigger the PWA router and navigate on the client side.

File: amp-pwa/src/components/amp-document.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+    /**
+     * XMLHTTPRequest that fetches the AMP document.
+     * @private
+     */
+    this.xhr_ = null;
+  }
+
+  componentDidMount() {
+    this.fetchAndAttachAMPDoc_(this.props.src);
```
Comment (dvoytenko) [inline amp-pwa/src/components/amp-document.js]: Generally, we do full camel-case style guide. E.g. `fetchAndAttachAmpDoc_`. If it doesn't conflict with React - let's switch.

File: amp-pwa/src/components/amp-document.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+export default class AMPDocument extends React.Component {
+  constructor(props) {
+    super(props);
+
+    // 'offline' is set to true if and when the document fetch fails.
+    this.state = {'offline': false};
+
+    /**
+     * `window.AMP` is set by the AMP runtime when it finishes loading.
+     * @private
```
Comment (dvoytenko) [inline amp-pwa/src/components/amp-document.js]: `@const`

File: amp-pwa/src/components/amp-document.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+     */
+    this.ampReadyPromise_ = new Promise(resolve => {
+      (window.AMP = window.AMP || []).push(resolve);
+    });
+
+    /**
+     * Child element that the AMP document will be added as a shadow root to.
+     * @private
```
Comment (dvoytenko) [inline amp-pwa/src/components/amp-document.js]: Please define type

File: amp-pwa/src/components/amp-document.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+     */
+    this.container_ = null;
+
+    /**
+     * XMLHTTPRequest that fetches the AMP document.
+     * @private
```
Comment (dvoytenko) [inline amp-pwa/src/components/amp-document.js]: type

File: amp-pwa/src/components/amp-document.js [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    }).catch(error => {
+      this.setState({'offline': true});
+    });
+  }
+
+  /**
+   * Fetches and parses HTML at `url`.
+   * @private
+   * @param {string} url
+   * @return {Promise} If fetch succeeds, resolved with {Document}.
```
Comment (dvoytenko) [inline amp-pwa/src/components/amp-document.js]: Should be `{!Promise<!Document>}`

File: amp-pwa/src/components/amp-document.js [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+   *         Otherwise, rejects with {string} error description.
+   */
+  fetchDocument_(url) {
+    return new Promise((resolve, reject) => {
+      const xhr = (this.xhr_ = new XMLHttpRequest());
```
Comment (dvoytenko) [inline amp-pwa/src/components/amp-document.js]: Looks like we also need to reset `this.xhr_ = null` in the either outcome of this promise.

File: amp-pwa/src/components/amp-document.js (hunk 4 above)
Comment (choumx) [inline amp-pwa/src/components/amp-document.js]: Done.

File: amp-pwa/src/components/amp-document.js (hunk 5 above)
Comment (choumx) [inline amp-pwa/src/components/amp-document.js]: Done.

File: amp-pwa/src/components/amp-document.js (hunk 6 above)
Comment (choumx) [inline amp-pwa/src/components/amp-document.js]: Done.

File: amp-pwa/src/components/amp-document.js (hunk 3 above)
Comment (choumx) [inline amp-pwa/src/components/amp-document.js]: Done.

File: amp-pwa/src/components/amp-document.js (hunk 7 above)
Comment (choumx) [inline amp-pwa/src/components/amp-document.js]: Done.

File: amp-pwa/src/components/amp-document.js (hunk 8 above)
Comment (choumx) [inline amp-pwa/src/components/amp-document.js]: Done.

**Comment by choumx** [comment]
Added cross-ref links to AMP documents and corresponding react-router support in `amp-document.js`.

Haven't updated the live demo with this yet -- ironically, prod AMP works for this feature but HEAD has issues.

**Comment by dvoytenko** [comment]
I'm good here. How about everyone else?

**Comment by choumx** [comment]
@cramforce Any other feedback?

**Pull request merged**